modules/staking_pool/src/lib.rs

```diff
@@ -239,56 +208,120 @@ impl<T: Trait> Module<T> {
 			.saturating_mul_int(amount)
 	}
 
-	pub fn redeem_by_free_unbonded(who: &T::AccountId, amount: Balance) -> DispatchResult {
-		let current_era = Self::current_era();
-		let mut total_deduct = amount;
-		let mut fee = Self::calculate_claim_fee(total_deduct, current_era);
-		let mut ldot_to_redeem = total_deduct.saturating_sub(fee);
+	/// This function must to be called in `with_transaction_result` scope to
+	/// ensure atomic
+	pub fn redeem_by_unbond(who: &T::AccountId, amount: Balance) -> DispatchResult {
+		let mut liquid_amount_to_redeem = amount;
 		let liquid_exchange_rate = Self::liquid_exchange_rate();
-		let mut unbond_amount = liquid_exchange_rate.saturating_mul_int(ldot_to_redeem);
-		let free_unbonded = Self::free_unbonded();
-
-		if !unbond_amount.is_zero() && !free_unbonded.is_zero() {
-			if unbond_amount > free_unbonded {
-				// free_unbonded is not enough, re-calculate actual redeem ldot
-				let new_ldot_to_redeem = liquid_exchange_rate
+		let mut staking_amount_to_unbond = liquid_exchange_rate
+			.checked_mul_int(liquid_amount_to_redeem)
+			.ok_or(Error::<T>::Overflow)?;
+		let communal_bonded_staking_amount = Self::get_communal_bonded();
+
+		if !staking_amount_to_unbond.is_zero() && !communal_bonded_staking_amount.is_zero() {
+			// communal_bonded_staking_amount is not enough, re-calculate
+			if staking_amount_to_unbond > communal_bonded_staking_amount {
+				liquid_amount_to_redeem = liquid_exchange_rate
 					.reciprocal()
 					.unwrap_or_default()
-					.saturating_mul_int(free_unbonded);
+					.saturating_mul_int(communal_bonded_staking_amount);
+				staking_amount_to_unbond = liquid_exchange_rate
+					.checked_mul_int(liquid_amount_to_redeem)
+					.ok_or(Error::<T>::Overflow)?;
```

Comment:
Here `staking_amount_to_unbound` calculation result would be equal to `communal_bonded_staking_amount`, if I didn't miss something.